### PR TITLE
#45 logging API with Zap Implementation

### PR DESCRIPTION
**Pull request by stepanvalyavskiy**
#45 logging API, Zap Implementation but without `config file` option

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -56,7 +56,7 @@ run:
 linters-settings:
   dupl:
     # tokens count to trigger issue, 150 by default
-    threshold: 25
+    threshold: 150
```
Comment (g4s8) [inline .golangci.yml:59]: @stepanvalyavskiy why so big change here?

File: degitx.go [hunk 2]
```diff
@@ -9,6 +9,8 @@ import (
 	"log"
 	"net"
 
+	"cqfn.org/degitx/logging"
+
```
Comment (g4s8) [inline degitx.go:13]: @stepanvalyavskiy can we remove blank line here?

File: logging/context.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		EncodeDuration: zapcore.StringDurationEncoder,
+		EncodeCaller:   zapcore.ShortCallerEncoder,
+	}
+	consoleEncoder = zapcore.NewConsoleEncoder(consoleEncoderConfig)
+	jsonEncoderConfig := consoleEncoderConfig
+	jsonEncoderConfig.EncodeLevel = zapcore.CapitalLevelEncoder
+	jsonEncoder = zapcore.NewJSONEncoder(jsonEncoderConfig)
+}
+
+func InitLocator(loc string) {
```
Comment (g4s8) [inline logging/context.go:53]: @stepanvalyavskiy I'd change it to `Init()` and add a check if this function was called previously, and `panic()` if so.

File: logging/logger.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	"os"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type LogLevel int8
+
+const (
+	Debug LogLevel = 0
```
Comment (g4s8) [inline logging/logger.go:18]: @stepanvalyavskiy you can use `iota` syntax for levels https://github.com/golang/go/wiki/Iota

File: logging/logger.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+func (log *Logger) Warnf(template string, args ...interface{}) {
+	log.internalSugaredLogger.Warnf(template, args...)
+}
+
+func (log *Logger) Errorf(template string, args ...interface{}) {
+	log.internalSugaredLogger.Errorf(template, args...)
+}
+
+func openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error) {
+	sink, closeOut, err := zap.Open("./errors")
```
Comment (g4s8) [inline logging/logger.go:111]: @stepanvalyavskiy this filename should be configration parameter

File: logging/logger.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	errSink, _, err := zap.Open("./logInternalErrors")
```
Comment (g4s8) [inline logging/logger.go:115]: @stepanvalyavskiy what is the difference between `errors` and `logInternalErrors`? Can we use one file for both?

File: logging/context.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	highPriority = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool { //nolint:gochecknoglobals
+		return lvl >= zapcore.ErrorLevel && logLvl.Enabled(lvl)
+	})
+
+	lowPriority = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool { //nolint:gochecknoglobals
+		return lvl < zapcore.ErrorLevel && logLvl.Enabled(lvl)
+	})
+
+	consoleEncoder zapcore.Encoder //nolint:gochecknoglobals
+	jsonEncoder    zapcore.Encoder //nolint:gochecknoglobals
```
Comment (g4s8) [inline logging/context.go:29]: @stepanvalyavskiy to avoid so many global variables here, we can define internal private `struct`ure for base settings and assign it in `Init()` call:
```go
mux sync.Mutex

func Init(node Node) {
  mux.Lock()
  defer mux.Unlock()
  if log != nil {
    panic(errLogAlreadyInitialized)
  }
  log = &log{/*all params here*/}
}
```

File: logging/context.go (hunk 3 above)
Comment (g4s8) [reply logging/context.go:53]: @stepanvalyavskiy also, we may need more params later, not only node locator hex string, so I'd pass the whole `*locators.Node` here

**Review by g4s8 — CHANGES_REQUESTED** [review]
@stepanvalyavskiy thanks, please check my comments

File: logging/logger.go (hunk 4 above)
Comment (g4s8) [reply logging/logger.go:18]: @stepanvalyavskiy alos, you can use https://godoc.org/golang.org/x/tools/cmd/stringer standard tool to automatically generate `String()` method for log levels:
```go
//go:generate stringer -type=LogLevel
type LogLevel int8
```

File: .golangci.yml (hunk 1 above)
Comment (stepanvalyavskiy) [reply .golangci.yml:59]: It was wrongly configured as 15, but default is 150.
25 triggers even 1 live delegating in logger functions.
Let set to 30, but I afraid we will need to increase it later once again.

File: degitx.go (hunk 2 above)
Comment (stepanvalyavskiy) [reply degitx.go:13]: sure, I'm surprised that `gofumpt` didn't do that or report.

File: logging/logger.go (hunk 4 above)
Comment (stepanvalyavskiy) [reply logging/logger.go:18]: > @stepanvalyavskiy you can use `iota` syntax for levels https://github.com/golang/go/wiki/Iota

I thought we have a linter to avoid iota, at least I've read an article that speaks against iota syntax. Can't find now.

Comment (stepanvalyavskiy) [reply logging/logger.go:18]: > @stepanvalyavskiy alos, you can use https://godoc.org/golang.org/x/tools/cmd/stringer standard tool to automatically generate `String()` method for log levels:
> 
> ```go
> //go:generate stringer -type=LogLevel
> type LogLevel int8
> ```

ok, done

File: logging/logger.go (hunk 6 above)
Comment (stepanvalyavskiy) [reply logging/logger.go:115]: yes we can, but `logInternalErrors` is specific for Zap logger. `errSink` is an output for internals zap errors, that wasn't logged by users of this lib.

File: .golangci.yml (hunk 1 above)
Comment (g4s8) [reply .golangci.yml]: @stepanvalyavskiy I see, then it's ok to have 150 as default value

**Review by g4s8 — CHANGES_REQUESTED** [review]
@stepanvalyavskiy one more comment: can you please start each commit with ticket number, e.g. `#45 - something changed` and rebase+forcepush to this branch again?

**Comment by stepanvalyavskiy** [comment]
> @stepanvalyavskiy one more comment: can you please start each commit with ticket number, e.g. `#45 - something changed` and rebase+forcepush to this branch again?

@g4s8 done

File: locators/locator.go [hunk 8]
```diff
@@ -22,6 +23,7 @@ type Node struct {
 	PubKey  []byte
 	PrivKey []byte
 	Addr    ma.Multiaddr
+	Mux     sync.Mutex
```
Comment (g4s8) [inline locators/locator.go:26]: @stepanvalyavskiy why mutex is in `locators` package? I think it'd be more logically to put it into logging `package`.

File: logging/context.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	"go.uber.org/zap/zapcore"
+)
+
+type LogConfig struct {
+	Outputs   []Output `yaml:"outputs"`
+	ErrorsOut []string `yaml:"errors"`
+}
+
+type Output struct {
+	UniqName string   `yaml:"u_name"`
```
Comment (g4s8) [inline logging/context.go:21]: @stepanvalyavskiy do we really need this name field? Maybe we can safely remove it?

**Review by g4s8 — CHANGES_REQUESTED** [review]
@stepanvalyavskiy thanks, please check my comments

File: logging/context.go (hunk 9 above)
Comment (stepanvalyavskiy) [reply logging/context.go:21]: We have a logger with multicores, so it's one logger that has different log level for several output groups.
For ex `DEBUG` for `stdout` & `file path 2` 
and `WARN` for `file path 2` & `some url`
If we want to change a log level only for one output group, it's impossible to `log.SetLevel($LVL)`, we need to ident output grpoup at first. Otherwise all output groups will became same log level after calling `log.SetLevel($LVL)`
To change log level only for one group I use  `log.SetLevel($LVL, $OUTPUT_UID)`

TLDR: If we remove this field, we lose an opportunity to change log level only for some output groups.

Comment (g4s8) [reply logging/context.go:21]: @stepanvalyavskiy what is the output group? I think we can initialize log from config only once, so we don't need `SetLogLevel()` method at all: we can just parse the config, and construct logger for each item from logging configuration list

Comment (stepanvalyavskiy) [reply logging/context.go:21]: > @stepanvalyavskiy what is the output group? I think we can initialize log from config only once, so we don't need `SetLogLevel()` method at all: we can just parse the config, and construct logger for each item from logging configuration list

then we can remove SetLevel and u_name as well.

**Review by g4s8 — APPROVED** [review]
@stepanvalyavskiy thanks!
